### PR TITLE
Fix Forta based monitoring

### PR DESCRIPTION
**Pull request by dgusakov**
https://www.notion.so/Fix-Forta-based-monitoring-48d03fd27424412b833f827511dba18c

Published to test agent https://explorer.forta.network/agent/0x219fcef9d4e4c040b1dd2feb61d5513e6957420b5f25d690e39a78004330af86

### CODE REVIEW

**Review by vgorkavenko — APPROVED** [review]

**Review by skhomuti — APPROVED** [review]

File: src/agent.ts [hunk 1]
```diff
@@ -81,60 +105,100 @@ const initialize = async (blockNumber: number, findings: Finding[]) => {
...
+          }
+          success = true
+        } catch (err) {
+          lastError = err
         }
-      } catch (err) {
-        findings.push(errorToFinding(err, agent, 'handleBlock'))
+      }
+      if (!success) {
+        blockFindingsCache.push(errorToFinding(lastError, agent, 'handleTransaction'))
```
Comment (arwer13) [inline src/agent.ts:146]: 'handleTransaction' --> 'handleBlock'

**Review by arwer13 — CHANGES_REQUESTED** [review]

File: src/agent.ts [hunk 2]
```diff
@@ -35,18 +40,42 @@ const subAgents: SubAgent[] = [
   agentEasyTrack,
 ]
 
-let initialized = false
-let initializedPromise: Promise<void> | null = null
+// block or tx handlinig should take no more than 5 sec. If not all processing is done it will be done later in background
+const handlerResolveTimeot = 5000
```
[the remark this answers is not in the document]
Comment (dgusakov) [reply src/agent.ts]: Done

File: src/agent.ts [hunk 3]
```diff
@@ -81,60 +105,100 @@ const initialize = async (blockNumber: number, findings: Finding[]) => {
...
-  let findings: Finding[] = []
 
-  await initializeIfNeeded(blockEvent.blockNumber, findings)
+  let responseResolve: (value: Finding[]) => void
+
+  const response = new Promise<Finding[]>((resolve,reject)=>{
+    responseResolve = resolve
+  });
+
+  // we need to resolve Promise in handlerResolveTimeot maximum.
```
[the remark this answers is not in the document]
Comment (dgusakov) [reply src/agent.ts]: Done

File: src/agent.ts [hunk 4]
```diff
@@ -81,60 +105,100 @@ const initialize = async (blockNumber: number, findings: Finding[]) => {
...
+  // If not all handlers has finished execution we will left them working in background
```
[the remark this answers is not in the document]
Comment (dgusakov) [reply src/agent.ts]: Done

File: src/agent.ts [hunk 5]
```diff
@@ -81,60 +105,100 @@ const initialize = async (blockNumber: number, findings: Finding[]) => {
...
+        txFindingsCache.push(errorToFinding(lastError, agent, 'handleTransaction'))
       }
     }
-  }))
-
-  return findings
-}
-
+  })).then(() => {
+    // if all hadlers have finished execution drop timeout and resolve promise
```
[the remark this answers is not in the document]
Comment (dgusakov) [reply src/agent.ts]: Done

File: src/agent.ts (hunk 1 above)
Comment (dgusakov) [reply src/agent.ts]: Done

**Comment by dgusakov** [comment]
@arwer13 All fixed. And a bit more. Now i have a spell checker plugin)

**Review by arwer13 — APPROVED** [review]

**Pull request merged**